stream_alert/alert_processor/main.py

```diff
@@ -46,350 +42,87 @@ def handler(event, context):
-            )
-        )
-        logger.info('Alert sent to S3!')
+        # Retrieve the proper class to handle dispatching the alerts of this services
+        output_dispatcher = get_output_dispatcher(service, region, function)
 
-    # TODO(jacknagz): investigate message order bug
-    def _slack(self, rule_name, alerts):
-        """Send alert text to Slack
+        if not output_dispatcher:
+            continue
 
-        Args:
-            rule_name: The name of the triggered rule.
-            alerts: The array of alerts relevant to the triggered rule.
-        """
-        baseurl = self.creds.get('slack').url
-        token = self.creds.get('slack').secret
-        url = os.path.join(baseurl, token)
+        try:
+            output_dispatcher.dispatch(descriptor, rule_name, alert)
+        except BaseException as err:
+            LOGGER.error('an error occurred while sending alert to %s: %s',
+                         service, err)
 
-        attachment = {
-            'color': 'danger',
-            'fallback': 'StreamAlert Rule Triggered - {}'.format(rule_name),
-            'author_name': 'StreamAlert Rule Triggered',
-            'ts': calendar.timegm(time.gmtime())
-        }
-        fields = [
-            {
-                'title': 'Rule',
-                'value': rule_name,
-                'short': True
-            },
-            {
-                'title': '# of Alerts',
-                'value': len(alerts),
-                'short': True
-            },
-            {
-                'title': 'Service',
-                'value': alerts[0]['metadata']['source']['service'],
-                'short': True
-            },
-            {
-                'title': 'Entity',
-                'value': alerts[0]['metadata']['source']['entity'],
-                'short': True
-            },
-        ]
+def sort_dict(unordered_dict):
+    """Recursively sort a dictionary
 
-        attachment['fields'] = fields
-        json_data = json.dumps({'attachments': [attachment]})
-        resp = self.request_helper(url, json_data)
+    Args:
+        unordered_dict [dict]: an alert dictionary
 
-        if resp and resp.read() == 'ok':
-            logger.info('Successfully sent attachment to Slack.')
-            success_cnt = 0
-            for alert in alerts:
-                text_data = json.dumps({'text': '```{}```'.format(
-                    json.dumps(alert['record'], indent=4)
-                )})
-                resp = self.request_helper(url, text_data)
-                if resp and resp.read() == 'ok':
-                    success_cnt += 1
-            logger.info('Successfully sent %s alerts to Slack.', success_cnt)
+    Returns:
+        [OrderedDict] a sorted version of the dictionary
+    """
+    result = OrderedDict()
+    for key, value in sorted(unordered_dict.items(), key=lambda t: t[0]):
+        if isinstance(value, dict):
+            result[key] = sort_dict(value)
+            continue
```

Comment:
An `else` block makes more sense than `continue` to me, but either way is fine. The value is either set recursively, or it is not